teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1071 @@
+<tr>
+<td>Example StepAction</td>
+<td>Example Task with orchestration in Steps</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step1
+spec:
+  image: alpine
+  script: |
+    echo "I am expendable"
+
+```
+</td>
+<td rowspan=2>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      stepRef:
+        name: step1
+      onError: continue
+    - name: s2
+      stepRef:
+        name: step2
+      computeResources:
+        requests:
+          memory: 1Gi
+          cpu: 500m
+        limits:
+          memory: 2Gi
+          cpu: 800m
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step2
+spec:
+  image: alpine
+  script: |
+    echo "I am special"
+  securityContext:
+    privileged: true
+```
+</td>
+</tr>
+</table>
+
+#### Params
+If a `StepAction` needs `Parameters`, they need to be declared so that they can be provided by the `Task` author. Without declaring `Parameters`, the `Task` author would have had to go through the `StepAction` specification to decipher what is needed. A `Task` should declare the `Parameters` that the underlying `Step` needs so that they can be provided by the `TaskRun` or the `Pipeline`. 
+<table>
+<tr>
+<td>
+
+```yaml
+# This is how a step author would 
+# declare the params that it needs.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: foo
+      type: string
+      description: “This is a param”
+      default: bar
+  image: alpine
+  script: |
+    echo $(params.foo)
+
+```
+</td>
+<td>
+
+```yaml
+# This is how a task author would 
+# provide the params that the step needs.
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  params:
+    - name: foo
+      type: string
+      description: “This is a param”
+      default: bar
+  steps:
+    - name: step1
+      stepRef: 
+        name: myStep
+```
+</td>
+</tr>
+</table>
+
+Referencing `StepAction` in a `Task` could create name conflicts that need to be resolved by the `Task` author. For example, if “step1” needs a `Parameter` called “url” and “step2” also needs a `Parameter` called “url” but they are actually meant for different purposes, the `Task` author needs to be able to pass in different values to both these `Parameters`.
+
+
+All inlined `Steps` and `Sidecars` in a `Task` can access all the `Parameters` and reference them in their work. This behavior is still intact. However, `Parameters` need to be passed explicitly to a referenced `Step`. While users only need to provide values when there are name conflicts (since a `Step` has access to all params of a `Task`) we start by explicitly providing all the required parameters to the `StepAction` and will be enforced during validation. To prevent increased verbosity of the specifications, we can invoke parameter propagation in future work. 
```

Comment:
I like the idea of introducing propagated params from the start. That way, users will just need to provide params where they need to resolve conflicts.

WDYT @jerop ?

Reply:
On second thoughts, I think we rejected propagated params for referenced Tasks and Pipelines because the usage of those parameters in the referenced Tasks/Pipelines is opaque to the users providing them. We only propagated params for inlined specification. For the same reason, I think we should not propagate params here since all the StepActions are referenced. 

To reduce scope here, we can revisit propagation of params and workspaces in referenced specification as a part of future work if needed. WDYT?

Reply:
As @wlynch pointed out as well that things like Context variables will also likely help reduce the verbosity since StepActions could potentially use them directly.

I think there was general agreement that we can scope this as future work.

Resolving this for now.